display_list/display_list_image_filter.h

```diff
@@ -665,6 +672,70 @@ class DlColorFilterImageFilter final : public DlImageFilter {
   std::shared_ptr<DlColorFilter> color_filter_;
 };
 
+class DlLocalMatrixImageFilter final : public DlImageFilter {
+ public:
+  explicit DlLocalMatrixImageFilter(const SkMatrix& matrix,
+                                    std::shared_ptr<DlImageFilter> filter)
+      : matrix_(matrix), image_filter_(filter) {}
+  explicit DlLocalMatrixImageFilter(const DlLocalMatrixImageFilter* filter)
+      : DlLocalMatrixImageFilter(filter->matrix_, filter->image_filter_) {}
+  DlLocalMatrixImageFilter(const DlLocalMatrixImageFilter& filter)
+      : DlLocalMatrixImageFilter(&filter) {}
+
+  std::shared_ptr<DlImageFilter> shared() const override {
+    return std::make_shared<DlLocalMatrixImageFilter>(this);
+  }
+
+  DlImageFilterType type() const override {
+    return DlImageFilterType::kLocalMatrixFilter;
+  }
+  size_t size() const override { return sizeof(*this); }
+
+  const SkMatrix& matrix() const { return matrix_; }
+
+  const DlLocalMatrixImageFilter* asLocalMatrix() const override {
+    return this;
+  }
+
+  bool modifies_transparent_black() const override { return false; }
+
+  SkRect* map_local_bounds(const SkRect& input_bounds,
+                           SkRect& output_bounds) const override {
+    output_bounds = matrix_.mapRect(input_bounds);
+    return &output_bounds;
+  }
+
+  SkIRect* map_device_bounds(const SkIRect& input_bounds,
+                             const SkMatrix& ctm,
+                             SkIRect& output_bounds) const override {
+    return image_filter_->map_device_bounds(
+        input_bounds, SkMatrix::Concat(ctm, matrix_), output_bounds);
```

Comment:
The input bounds are relative to `ctm`, but not to `matrix_`. You need to return bounds relative to *only* `ctm`.

Reply:
Actually, I'm wondering what Skia does with this. Have you run tests to see how the bounds methods behave with a local matrix?

Reply:
I added a test case:
```C++
auto blur_filter = SkImageFilters::Blur(5.0, 6.0, SkTileMode::kRepeat, nullptr);
  auto local_filter = blur_filter->makeWithLocalMatrix(SkMatrix::Scale(2, 2));
  auto inputBounds = SkIRect::MakeLTRB(20, 20, 80, 80);
  auto rect = local_filter->filterBounds(inputBounds, SkMatrix::I(), SkImageFilter::MapDirection::kForward_MapDirection);

  auto dl_color_filter = std::make_shared<DlBlurImageFilter>(5.0, 6.0, DlTileMode::kRepeat);
  auto local_matrix_filter = DlLocalMatrixImageFilter(SkMatrix::Scale(2, 2), dl_color_filter);
  SkIRect out_bounds;
  local_matrix_filter.map_device_bounds(inputBounds, SkMatrix::I(),out_bounds);
  ASSERT_EQ(out_bounds, rect);
```
the out_bouns is equal the rect when we use `SkMatrix::Concat(ctm, matrix_)`

Reply:
This is a very basic test of some of the cases that have the fewest quirks in them.

Try something like:

```
For each filter type:
  filter = instantiate that type;
  For each transform type local_matrix:
    transformed_filter = filter->makeWithLocal(local_matrix)
    For each transform type map_matrix:
      compare output of transformed_filter.map(..., map_matrix, ...);
```
Transform types should be:
  - translate
  - scale + translate
  - rotate + translate
  - perspective
  - any others?